hadoop-hdds/docs/content/design/container-reconciliation.md

```diff
@@ -0,0 +1,344 @@
++ */
++message ContainerMerkleTree {
++  enum FailureCause {
++    NO_HEALTHY_CHUNK_FOUND_WITH_PEERS = 1; // No healthy chunk found with peers.
++    NO_PEER_FOUND = 2; // No peer found.
++  }
++  optional int64 containerID = 1; // The container ID.
++  repeated BlockMerkleTree blockMerkleTrees = 2; // The blocks in this should be sorted by the order of blocks written.
++  optional ChecksumData checksumData = 3; // Checksum of the checksums of the blocks.
++  optional int64 length = 5; // Length of the container.
++  optional int64 blockCount = 6; // Number of blocks in the container.
++  optional FailureCause failureCause = 7; // The cause of the failure.
++  optional int64 reconciliationCount = 8; // The reconciliation count.
++}
+
+
+```
+This is written as a file to avoid bloating the RocksDB instance which is in the IO path.
+
+## APIs
+
+The following APIs would be added to datanodes to support container reconciliation. The actions performed when calling them is defined in [Events](#events).
+
+- `reconcileContainer(containerID, List<Replica>)`
+    - Instructs a datanode to reconcile its copy of the specified container with the provided list of other container replicas.
+    - Datanodes will call `getContainerHashes` for each container replica to identify repairs needed, and use existing chunk read/write APIs to do repairs necessary.
+    - This would likely be a new command as part of the SCM heartbeat protocol, not actually a new API.
+- `getContainerHashes(containerID)`
+    - A datanode API that returns the merkle tree for a given container. The proto structure would be similar to that outlined in [Merkle Tree](#Container-Hash-Tree-/-Merkle-Tree).
+
+## Reconciliation Process
+
+SCM: Storage Container Manager
+
+DN: Datanode
+
+### SCM sets up the reconciliation process as follows:
+
+1. SCM triggers a reconciliation on `DN 1` for container 12 with replicas on `DN 2` and `DN 3`.
+   1. `SCM -> reconcileContainer(Container #12, DN2, DN3) -> DN 1` 
+
+### Datanodes set up the reconciliation process as follows:
+
+1. `DN 1` schedules the Merkle Tree to be calculated if it is not already present and report it to SCM. 
+   1. SCM can then compare the container replica hashes and schedule reconciliation if they are different.
+2. If `DN 1` already has the Merkle Tree locally, it will compare it with the Merkle Trees of the other container replicas and schedule reconciliation if they are different. Example: 
+   1. `DN 1 -> getContainerHashes(Container #12) -> DN 2` // Datanode 1 gets the merkle tree of container 12 from Datanode 2.
+   2. `DN 1 -> getContainerHashes(Container #12) -> DN 3` // Datanode 1 gets the merkle tree of container 12 from Datanode 3.
+   3. ... // Continue for all replicas.
+
+### Reconcile loop once the merkle trees are obtained from all/most replicas:
+
+1. `DN 1` checks if any blocks are missing. For each missing Block:
```

Comment:
What's the criteria to determine whether a block is missing, or merely an extra not needed block?  The question is what's the rule to decide which replica is the baseline replica.

Reply:
1. The reconciliation process does not take into account if a block is referenced or not by OM. If the block was written to a container it will be replicated across as a safety measure. Recon can at a later point identify unwanted blocks.
2. Yes when adding a block the metadata for it will also need to be updated.